xbean-finder/src/main/java/org/apache/xbean/finder/archive/JarArchive.java

```diff
@@ -41,22 +44,63 @@ public class JarArchive implements Archive, AutoCloseable {
     private final MJarSupport mjar = new MJarSupport();
 
-    public JarArchive(ClassLoader loader, URL url) {
+    /*
+     * Supports only 'file:/...' or 'jar:file:/...!/' URLs
+     */
+    public JarArchive(ClassLoader loader, URL url){
 //        if (!"jar".equals(url.getProtocol())) throw new IllegalArgumentException("not a jar url: " + url);
 
-        try {
-            this.loader = loader;
-            this.url = url;
-            URL u = url;
+        this.loader = loader;
+        this.url = url;
+        File jarFile = null;
+        String jarPath;
+        int idx;
 
-            String jarPath = url.getFile();
-            if (jarPath.contains("!")) {
-                jarPath = jarPath.substring(0, jarPath.indexOf("!"));
-                u = new URL(jarPath);
+        // Wipe out 'jar:' prefix AND '!/{...}' suffix(if any)
+        if("jar".equalsIgnoreCase(url.getProtocol())){
+
+            try{
+                jarPath = url.getPath();
+                url = new URL(jarPath.substring(0, jarPath.lastIndexOf("!/")));
+            }catch(MalformedURLException ex){
+                // Probably CPU overheat and/or DRAM undervoltage
+                throw new UnsupportedOperationException(
+                        "Please provide 'file:/...' or 'jar:file:/...!/' URL"
+                                + " instead of '" + FileArchive.decode(String.valueOf(url)) + "'");
             }
-            jar = new JarFile(FileArchive.decode(u.getFile())); // no more an url
-        } catch (IOException e) {
-            throw new IllegalStateException(e);
+        }
+
+        try{
+            // handle 'file:/...' URL
+            if("file".equalsIgnoreCase(url.getProtocol())){
+
+                // Testing if file DOEN't exists AND cutting !/{...}'
+                //  suffix-by-suffix until file discovered or run out of suffixes
+                for(jarPath = FileArchive.decode(url.getPath());
+                        !(jarFile = new File(jarPath)).exists()
+                            && (idx = jarPath.lastIndexOf("!/")) > 0;
+                        jarPath = jarPath.substring(0, idx)){ }
+
+                // No more suffixes to cut, but referenced file wasn't discovered
+                if(!jarFile.exists()){
+
+                    // To be caught later and wrapped into IllegalStateEx - default behavior
+                    throw new FileNotFoundException(FileArchive.decode(String.valueOf(url)));
+                }
+
+            }else{
+                throw new UnsupportedOperationException(
+                        "Please provide 'file:/...' or 'jar:file:/...!/' URL"
+                                + " instead of '" + FileArchive.decode(String.valueOf(url)) + "'");
+            }
+
+            jar = new JarFile(jarFile);
+
+        }catch(IOException e){
+            throw new IllegalStateException("Cannot open jar(zip) '"
+                    + jarFile != null ? // why can it be null? but since compiler thinks so...
```

Comment:
Can if none of protocols match so no compiler issue ;)

Reply:
ok, it knows better